enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,592 @@
+  metal specific tasks that should be declared outside of the scope of
+  the MAO, particularly since it does not have equivalent
+  responsibilities on any other platform.
+
+* Expanding needs for bare metal specific manifests - for example, new
+  CRDs used to drive new baremetal capabilities - to be installed
+  early in the cluster bring-up means introducing yet more bare metal
+  specific concerns into the MAO.
+
+Steps (2) and (3) are aspects of cluster bring-up which the CVO is
+clearly well-suited. However, to date, it was understood that creating
+a Second Level Operator (SLO) (in other words, a CVO-managed operator)
+for bare metal would not make sense, since it implied a component
+installed and running on clusters where it is not needed.
+
+### Goals
+
+Allow bare metal machine management capabilities to be fully
+encapsulated in a new SLO.
+
+Ensure that this new SLO has minimal impact on clusters not running on
+the `baremetal` platform.
+
+### Non-Goals
+
+### Proposal
+
+Recognizing that bare metal support warrants the creation of a new
+"subsystem":
+
+1. Create a new, OpenShift-specific project called
+   `cluster-baremetal-operator` aka CBO.
+2. This project should implement a new SLO which is responsible for
+   installing bare metal specific CRDs and running the BareMetalHost
+   controller and related Metal3 components.
+3. The CBO should meet all of the standard expectations of an SLO,
+   including for example keeping a ClusterOperator resource updated.
+4. On infrastructure platforms other than `baremetal`, the CBO should
+   following emerging patterns for "not in use" components, for
+   example setting its cluster operator condition to `Disabled=true`,
+   `Available=true`, and `Progressing=False`.
+
+### Risks and Mitigations
+
+(FIXME)
+
+- Impact on non-baremetal platforms
+- Implementation cost and complexity
+- Upgrades
+- API stability
+- Additional bare metal specific functionality in future
+- Upstream (Metal3) implications
+
+## Design Details
+
+`cluster-baremetal-operator` is a new Second Level Operator (SLO)
+whose operand is a controller for the BareMetalHost resource and
+associated components from the Metal3 project. The below sections
+covers different areas of the design of this new SLO.
+
+### Standard SLO Behaviors
+
+As an SLO, the BMO is expected to adhere to the standard expected
+behaviours of SLO, including:
+
+1. The CBO image should be tagged with
+   `io.openshift.release.operator=true` and contain a `/manifests`
+   directory with all of the manifests it requires the CVO to apply,
+   along with an `image-references` file listing the images referenced
+   by those manifests that need to be included in the OpenShift
+   release image.
+2. Implement a `ClusterOperator` resource, including updating its
+   status field with the `Available/Progressing/Degraded` conditions,
+   operator version number, and any `relatedObjects`.
```

Comment:
Could we use these status fields like`Degraded=true` or `Available=false,Progressing=true` to relay information to the installer about how the underlying provisioning activities are going?  

The installer takes a naïve approach in determining failure: if it's taking longer than 30m (60m for metal) for all operators to come up, then it's marked as failed. The installer can't tell the difference between "this is taking a while" vs "this will never complete." 

However, it's conceivable to exceed 60m. POST can take a long time, or UEFI firmwares can be buggy and need to retry PXE.  Indeed, [we retry PXE on failure](https://github.com/metal3-io/ironic-image/commit/3fb8b0c62e767c0df88282b28b40ba2069fe96ec#diff-7fe4f348eb98d89bcbe2e453207d8d8cR55) and have seen need for it in some environments.

I'd like to see us take a more intelligent approach for baremetal. If we know we're still working on provisioning and Metal3's Ironic hasn't given up, then we could conceivably wait longer. A recent, successful deployment at a cluster with 3 masters and 6 workers took 97m.

Alternatively, we might know 15m into the process that your BMC creds are garbage, and we could fail *early*.

It's a really poor UX to have the installer make users wait when it's knowable earlier that the install is failed, and similarly bad UX to fail when we know the underlying infrastructure is still making progress and there's a good chance of success.

I'm not even sure if that belongs in CBO or MAO, and maybe it belongs as a separate enhancement request.

What do you think?

Reply:
It's a very good suggestion. Honestly, we could probably trust ironic to apply its own timeouts (and adjust ironic-image as we see fit).